### PR TITLE
API: resolve unset vmin / vmax in all ScalarMapple based methods

### LINKED ISSUE #17703: QuadMesh.get_clim changed behavior in 3.3.0rc1
### Bug report

**Bug summary**

`QuadMesh.get_clim()` is documented as:
> Return the values (min, max) that are mapped to the colormap limits.

Without `vmin` or `vmax` arguments to `pcolormesh`, this returns the data minimal and maximal values. However in the presence of only one of `vmin` or `vmax`, a change occurred between 3.2.2 and 3.3.0rc1. In 3.3.0rc1 it returns the input vmin *and* vmax, though only one of them was specified. The provided examples also happens (with slight output changes) for setting `vmax` instead of vmin.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

dx, dy = 0.15, 0.05
y, x = np.mgrid[slice(-3, 3 + dy, dy),
                slice(-3, 3 + dx, dx)]
z = (1 - x / 2. + x ** 5 + y ** 3) * np.exp(-x ** 2 - y ** 2)
z = z[:-1, :-1]

f, ax = plt.subplots(1,1)
ax.pcolormesh(x, y, z, vmin=0)
ax.axis([x.min(), x.max(), y.min(), y.max()])
for mesh in ax.collections:
    print(mesh.get_clim())
```

**Actual outcome**

```
# 3.3.0rc1
(0.0, None)
```

**Expected outcome**
```
# 3.2.2
(0.0, 1.0510083319982622)
```

**Matplotlib version**
  * Operating system: MacOSX
  * Matplotlib version: 3.3.0rc1
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX
  * Python version: 3.8.3

Installed using pip

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

See analysis in #17703 

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

- Resolves #17703

### CODE REVIEW

File: doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst [hunk 1]
```diff
@@ -312,3 +312,30 @@ rcParam is True.
...
+first draw if only one of the *vmin* or *vmax* keyword arguments was
+passed (`.Axes.scatter`, `.Axes.hexbin`, `.Axes.imshow`,
+`.Axes.pcolorfast`) but would scale based on the passed data if
+neither was passed (independent if the  *norm* kwarg).
+
+Other  methods (`.Axes.pcolor`, `.Axes.pcolormesh`)
+always autoscaled base on the initial data.
+
+All of the plotting methods now resolve, based on the passed data, the
+the unset *vmin* or *vmax* at init time based on the data passed.
```
Comment (timhoffm) [inline doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst:330]:
```suggestion
All of the plotting methods now resolve an unset *vmin* or *vmax* at
init time based on the data passed.
```
Or similar. The original sentence seems broken.

Comment (timhoffm) [reply doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst:330]: Travis OSX CI is currently broken and the failure is unrelated to this PR.

File: doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst [hunk 2]
```diff
@@ -312,3 +312,30 @@ rcParam is True.
...
+
+Aggressively autoscale clim in ``ScalerMappable`` classes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+Previously some plotting methods would defer autoscaling until the
+first draw if only one of the *vmin* or *vmax* keyword arguments was
+passed (`.Axes.scatter`, `.Axes.hexbin`, `.Axes.imshow`,
+`.Axes.pcolorfast`) but would scale based on the passed data if
+neither was passed (independent if the  *norm* kwarg).
```
Comment (QuLogic) [inline doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst:324]:
```suggestion
neither was passed (independent of the  *norm* keyword argument).
```

File: doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst [hunk 3]
```diff
@@ -312,3 +312,30 @@ rcParam is True.
...
+If you were relying on exactly one of the *vmin* or *vmax* remaining
+unset between the time when the method is called and the first time
+the figure is rendered you get back the old behavior by manually setting
+the relevant limit back to `None` ::
+
+  cm_obj.norm.vmin = None
+  # or
+  cm_obj.norm.vmax = None
+
+which will be resolved at during the draw process.
```
Comment (QuLogic) [inline doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst:341]:
```suggestion
which will be resolved during the draw process.
```

File: doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst [hunk 4]
```diff
@@ -312,3 +313,29 @@ rcParam is True.
...
+
+Previously some plotting methods would defer autoscaling until the
+first draw if only one of the *vmin* or *vmax* keyword arguments were
+passed (`.Axes.scatter`, `.Axes.hexbin`, `.Axes.imshow`,
+`.Axes.pcolorfast`) but would scale based on the passed data if
+neither was passed (independent of the *norm* keyword arguments).
+Other methods (`.Axes.pcolor`, `.Axes.pcolormesh`) always autoscaled
+base on the initial data.
+
+All of the plotting methods now resolve the the unset *vmin* or *vmax*
```
Comment (efiring) [inline doc/api/prev_api_changes/api_changes_3.3.0/behaviour.rst:329]: "the the"

**Pull request merged**